File: backend/src/routes/stats.js
import express from 'express';
import { prisma } from '../db.js';
import { get as cacheGet, set as cacheSet } from '../cache.js';

const router = express.Router();

/** GET /api/stats — totals, sums, breakdowns, last fetch time */
router.get('/', async (_req, res) => {
  try {
    const cached = cacheGet('stats');
    if (cached) {
      return res.json(cached);
    }

    const openWhere = {
      status: 'open',
      OR: [
        { endDate: { gte: new Date() } },
        { endDate: null }
      ]
    };

    const now = new Date();
    const [totalOpen, valueAgg, bySource, categoriesRaw, lastLog] = await Promise.all([
      prisma.tender.count({ where: openWhere }),
      prisma.tender.aggregate({ where: openWhere, _sum: { bidValue: true } }),
      prisma.tender.groupBy({ by: ['source'], where: openWhere, _count: { _all: true } }),
      prisma.tender.aggregateRaw({
        pipeline: [
          {
            $match: {
              status: 'open',
              $or: [
                { endDate: { $gte: { $date: now.toISOString() } } },
                { endDate: null }
              ]
            }
          },
          { $unwind: '$category' },
          { $group: { _id: '$category', count: { $sum: 1 } } }
        ]
      }),
      prisma.fetchLog.findFirst({ orderBy: { runAt: 'desc' } }),
    ]);

    const categoryCounts = {};
    if (Array.isArray(categoriesRaw)) {
      for (const item of categoriesRaw) {
        if (item && item._id) {
          let count = 0;
          if (typeof item.count === 'number') {
            count = item.count;
          } else if (item.count && item.count.$numberLong) {
            count = parseInt(item.count.$numberLong, 10);
          } else if (item.count && item.count.$numberInt) {
            count = parseInt(item.count.$numberInt, 10);
          } else if (item.count) {
            count = parseInt(String(item.count), 10) || 0;
          }
          categoryCounts[item._id] = count;
        }
      }
    }

    const result = {
      totalOpenTenders: totalOpen,
      totalEstimatedValue: valueAgg._sum.bidValue || 0,
      bySource: Object.fromEntries(bySource.map((s) => [s.source, s._count._all])),
      byCategory: categoryCounts,
      lastFetchAt: lastLog?.runAt || null,
    };

    cacheSet('stats', result, 3600000); // 1 hour

    res.json(result);
  } catch (e) {
    console.error('[api] GET /stats error:', e);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend/app/routes/stats.py
"""
app/routes/stats.py

Direct port of src/routes/stats.js.
"""
from datetime import datetime, timezone

from fastapi import APIRouter
from fastapi.responses import JSONResponse

from .. import db
from ..cache import get as cache_get, set as cache_set
from ..serialize import serialize_value

router = APIRouter(prefix="/api/stats", tags=["stats"])


@router.get("")
@router.get("/")
async def get_stats():
    """GET /api/stats — totals, sums, breakdowns, last fetch time"""
    try:
        cached = cache_get('stats')
        if cached:
            return JSONResponse(cached)

        now = datetime.now(timezone.utc)
        open_where = {'status': 'open', '$or': [{'endDate': {'$gte': now}}, {'endDate': None}]}

        col = db.tenders

        total_open = await col.count_documents(open_where)

        value_agg_cursor = col.aggregate([
            {'$match': open_where},
            {'$group': {'_id': None, 'sum': {'$sum': '$bidValue'}}},
        ])
        value_agg = await value_agg_cursor.to_list(length=1)
        total_estimated_value = value_agg[0]['sum'] if value_agg else 0

        by_source_cursor = col.aggregate([
            {'$match': open_where},
            {'$group': {'_id': '$source', 'count': {'$sum': 1}}},
        ])
        by_source_raw = await by_source_cursor.to_list(length=None)
        by_source = {s['_id']: s['count'] for s in by_source_raw}

        categories_raw = await col.aggregate([
            {'$match': open_where},
            {'$unwind': '$category'},
            {'$group': {'_id': '$category', 'count': {'$sum': 1}}},
        ]).to_list(length=None)
        category_counts = {item['_id']: item['count'] for item in categories_raw if item.get('_id')}

        last_log = await db.fetch_logs.find_one(sort=[('runAt', -1)])

        result = {
            'totalOpenTenders': total_open,
            'totalEstimatedValue': total_estimated_value or 0,
            'bySource': by_source,
            'byCategory': category_counts,
            'lastFetchAt': serialize_value(last_log['runAt']) if last_log else None,
        }

        cache_set('stats', result, 3600000)  # 1 hour

        return JSONResponse(result)
    except Exception as e:
        print(f"[api] GET /stats error: {e}")
        return JSONResponse({'error': 'Internal server error'}, status_code=500)
